src/main/java/com/google/edith/servlets/UserInsights.java

```diff
@@ -0,0 +1,194 @@
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * This class provides the funttionality to parse specifc user data from 
+ * datastore and send an agreggate of that information in a JSON file.
+ */
+public final class UserInsights {
+
+  private String userId;
+
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();  
+  private static final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+  private static final Gson gson = new Gson();
+
+  /** Assumes dates are in the form yyyy-MM-dd */
+  private static final Comparator<Key> SORT_BY_DATE = (Key item1, Key item2) -> {
+      try {
+        return ((String) (datastore.get(item1).getProperty("date")))
+            .compareTo((String) (datastore.get(item2).getProperty("date")));
+      } catch(EntityNotFoundException | NullPointerException e )  {
+        return 0;
+      }
+  };
+  
+  public UserInsights(String userId) {
+    this.userId = userId;
+  }
+
+  /**
+   * Finds the UserStats entity corresponding to {@code userId} in datastore.
+   * @return UserStats entity for this user
+   */
+  private Entity retreiveUserStats() {
+    Filter idFilter = new FilterPredicate("UserId", 
+                                FilterOperator.EQUAL,
+                                userId);
+    Query query = new Query("UserStats").setFilter(idFilter);
+    PreparedQuery results = datastore.prepare(query);
+    return results.asList(FetchOptions.Builder.withLimit(10)).get(0);
+  }
+
+  /** This should only be called each time a new user makes an accout. */
+  public void createUserStats() { 
+     List<Key> items = new ArrayList<>();
+
+     Entity userStats = new Entity("UserStats");
+     userStats.setProperty("UserId", userId);
+     userStats.setProperty("Items", items);
+     datastore.put(userStats);
+  }
+
+  /** 
+   * Updates the Items list in the UserStats Entity in datastore
+   * corresponding to this user.
+   * @param items - Non-null list of item Keys to be added to the current Item list.
+   */
+  public void updateUserStats(List<Key> newItems) {
+    Entity userStats = retreiveUserStats();    
+    List<Key> items = (List<Key>) userStats.getProperty("Items");
+    if (items == null) {
+      items = newItems;
+    } else {
+      items.addAll(newItems);
+    }
+    userStats.setProperty("Items", items);
+    datastore.put(userStats);
+  }
+ 
+  /** 
+   * Copmiles the spending using the Item list found in this user's
+   * UserStats Entity in datastore.
+   * TODO: Allow for various time periods (only calculates weekly aggregate now).
+   * @return A map relating a time period to the spending in that time period.
+   */
+  public Map<String, String> aggregateUserData() { 
```

Comment:
Typically we want to return as specific a type as possible while accepting as generic a type as possible in params.
This might be a great place to start using ImmutableMap as the return type. See
https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableMap.html

You could call ImmutableMap.copyOf. If this is too much overhead you can return a HashMap and pair program the ImmutableMap part later

Reply:
Done.